### PR TITLE
Tree shakeable v4.0 

### PR DESCRIPTION
**Pull request by jwagner**
### Release plan
- [x] Finish writing article
- [x] Finish demo
- [x] Release docs
- [ ] Merge this branch & Trigger NPM Release
- [ ] Release synth wave demo to 29a.ch
- [ ] Make synth wave demo repo public
- [ ] Fix up code pens
- [ ] Publish article & update on 29a.ch
- [ ] Announce new version (twitter, ?) 
- [ ] Clean up related issues & PRs

### CODE REVIEW

**Comment by redblobgames** [comment]
I very much want to try this! I was travelling and catching up now, haven't had a chance.

File: README.md [hunk 1]
```diff
@@ -22,59 +22,56 @@ Created something awesome with simplex-noise? Let me know so I can add it to the
...
-```javascript
-const simplex = new SimplexNoise('seed'),
-    value2d = simplex.noise2D(x, y),
-    sameSeed = new SimplexNoise('seed'),
-    differentSeed = new SimplexNoise('different seed');
+### 2D
 
-sameSeed.noise2D(x, y) === value2d
-differentSeed.noise2D(x, y) !== value2d
+```
```
Comment (mreinstein) [inline README.md:42]: nit: javascript syntax highlighting missing here

File: README.md [hunk 2]
```diff
@@ -22,59 +22,56 @@ Created something awesome with simplex-noise? Let me know so I can add it to the
...
-You can also pass an alternative random function to the constructor that is
-used to build the permutation table.
-This can be used with a custom pseudo random number generator:
+### 3D
 
-```javascript
-const random = new Alea(seed),
-    simplex = new SimplexNoise(random),
-    value2d = simplex.noise2D(x, y);
+```
```
Comment (mreinstein) [inline README.md:49]: nit: javascript syntax highlighting missing here

File: README.md [hunk 3]
```diff
@@ -22,59 +22,56 @@ Created something awesome with simplex-noise? Let me know so I can add it to the
...
+const noise3D = noiseFunction3D();
+console.log(noise3D(x, y, z));
 ```
 
-The ALEA PRNG can be found in the npm package [alea](https://npmjs.org/package/alea).
+### 4D
 
-## node.js
+```
```
Comment (mreinstein) [inline README.md:56]: nit: javascript syntax highlighting missing here

File: README.md [hunk 4]
```diff
@@ -22,59 +22,56 @@ Created something awesome with simplex-noise? Let me know so I can add it to the
...
+### Using a seed value
 
-```javascript
-const SimplexNoise = require('simplex-noise'),
-    simplex = new SimplexNoise(Math.random),
-    value2d = simplex.noise2D(x, y);
+By default simplex-noise.js will use Math.random() to seed the noise.
+You can pass in a PRNG function to use your own seed value.
+
+```
```
Comment (mreinstein) [inline README.md:66]: nit: javascript syntax highlighting missing here

File: README.md [hunk 5]
```diff
@@ -22,59 +22,56 @@ Created something awesome with simplex-noise? Let me know so I can add it to the
...
+import Alea from 'alea';
+const alea = new Alea('seed');
+const noise2D = noiseFunction2D(alea);
```
Comment (mreinstein) [inline README.md:69]: I don't like having the word `function` in the function name. Can we keep the same names we currently have? `noise2D`, `noise3D`, etc. ?

Comment (mreinstein) [reply README.md:69]:
```javascript
import { noise2D } from 'simplex-noise';

const noise = noise2D(new Alea('seed'));
```

**Review by mreinstein — CHANGES_REQUESTED** [review]
I would like to hear other opinions too, my only 2 issues are around the name of the api functions we expose, and eliminating that (accidental?) dependency `microtime` that snuck in. Other than that the rest of my comments are formatting nits

File: package.json [hunk 6]
```diff
@@ -77,5 +75,8 @@
     "docs": "typedoc --excludePrivate --out public/docs simplex-noise.ts",
     "prepare": "npm run-script build",
     "benchmark": "parcel build && node ./perf/benchmark.js"
+  },
+  "dependencies": {
+    "microtime": "^3.0.0"
```
[the remark this answers is not in the document]
Comment (jwagner) [reply package.json:80]: Good catch. I blindly followed the instructions for benchmark.js here: https://benchmarkjs.com/

File: README.md (hunk 5 above)
Comment (jwagner) [reply README.md:73]: I didn't want to call the factories/constructors noiseND to leave that name up to the user. If you extend the above example to also construct a 3d noise function you'd end up with something like noise2 = noise2D(), noise3 = noise3D() or n2 = noise2D(), n3 = noise3D(). It is also a bit more obvious to me from the name that noiseFunction2D returns a function, whereas noise2D I'd expect to be the noise function itself. It's a very weakly held opinion though.

Comment (mreinstein) [reply README.md:73]: users can choose whatever name they like for the import though. For example:

```javascript
import { noise2D as nz } from 'simplex-noise'

const noise2D = nz(new Alea('seed'))
```

Comment (mreinstein) [reply README.md:73]: another option is maybe `simpleNoiseXD` ?  

```javascript
import { simplexNoise2D } from 'simplex-noise'
```

**Comment by redblobgames** [comment]
I agree that noise2DFunction and noise2D are different enough that they shouldn't have the same name. Another possible name for noiseFunction: makeNoise2D or createNoise2D. That way instead of saying what it *is* ("Function") the name says what it *does* ("Make" or "Create" the "noise2D" function).

**Comment by redblobgames** [comment]
Ok, I tried it, I like it! With a bare bones program, minified but not gzipped:

- 5117 bytes with the non-treeshakeable library
- 1258 bytes with noise2D only
- 1697 bytes with noise3D only
- 2098 bytes with noise4D only
- 3597 with noise2D, noise3D, noise4D
- 5288 with noise2D, noise3D, noise4D, alea

I also tested with [mapgen4](https://www.redblobgames.com/maps/mapgen4/). The build went from 185619 bytes to 181892 bytes (before gzip).

Are you thinking of including a convenience wrapper as a default export?

```ts
export class SimplexNoise {
    constructor(random: RandomFn = Math.random) {
        this.noise2D = noiseFunction2D(random);
        this.noise3D = noiseFunction3D(random);
        this.noise4D = noiseFunction4D(random);
    }
}
export default SimplexNoise;
```

(not sure how to handle the alea part in a wrapper, as it means we'd have to import alea even when we're not using it; maybe that's ok with tree shaking?)

If you're not distributing alea, I think you can drop alea.md.

I feel like a verb `makeNoise2D()` (makeX suggests that it will return X) would look nicer than a noun `noiseFunction2D()`.

This new api looks so clean and pure! 🍀  Call the noise constructor, and you have a noise function. I can't imagine a simpler api.

File: README.md (hunk 5 above)
Comment (jwagner) [reply README.md:78]: Renamed them to createNoiseND() I hope that is a bit better for you as well. :)

**Comment by mreinstein** [comment]
LGTM, nice work @jwagner

**Comment by jwagner** [comment]
Thank you @mreinstein . :) I'm currently working on a little demo and possibly a tutorial to go along with 4.0 and then I think it's good to release.

**Comment by jwagner** [comment]
@redblobgames looks like using the pure annotation enables tree shaking without having to move the consts inside of the closure. Thanks for the challenge. :)

**Comment by mreinstein** [comment]
```javascript
// TODO: insert typescript related rant here
```

:)

**Comment by jwagner** [comment]
@mreinstein @redblobgames the release is done! Thanks a lot for your comments and support. It gave me to motivation needed to release this. I really appreciate it. :)

**Pull request merged**